File: Problem234/Solution.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	slow := head
	fast := head
	var stack []int
	for fast != nil && fast.Next != nil {
		stack = append(stack, slow.Val)
		fast = fast.Next.Next
		slow = slow.Next
	}
	if fast != nil {
		slow = slow.Next
	}
	i := len(stack) - 1
	for slow != nil {
		top := stack[i]
		if slow.Val != top {
			return false
		}
		slow = slow.Next
		i--
	}
	return true
}

File: Problem234/Solution.java
class Solution {
    public boolean isPalindrome(ListNode head) {
        Stack<Integer> stack = new Stack<>();

        ListNode slow = head;
        ListNode fast = head;
        while(fast.next != null && fast.next.next != null) {
            stack.add(slow.val);
            slow = slow.next;
            fast = fast.next.next;
        }

        if (fast.next != null) {
            stack.add(slow.val);
        }

        ListNode tail = slow.next;

        while (tail != null) {
            if (tail.val != stack.pop()) {
                return false;
            }
            tail = tail.next;
        }

        return stack.isEmpty();
    }
}